docs/proposals/event-notification.md

```diff
@@ -0,0 +1,143 @@
+
+**Devfile writer**: A devfile writer may be a runtime developer (for example, a Red-Hatter working on WildFly or an IBMer working on OpenLibery) creating a devfile for their organization's runtime (for example 'OpenLiberty w/ Maven' dev file), or an application developer creating/customizing a dev file for use with their own application. In either case, the devfile writer must be familiar with the semantics of both odo and the devfile.
+
+## JSON-based odo command behaviours to detect app and build status
+
+New odo commands and flags:
+- `odo push -o json`
+- `odo component status -o json`
+
+With these two additions, an IDE or similar external tool can detect build running/succeeded/failed, application starting/started/stopping/stopped, and (in many cases) get a 'detailed app status' and/or 'detailed build status'.
+
+### Build status notification via `odo push -o json`
+
+`odo push -o json` is like standard `odo push`, but instead it outputs JSON events (including action console output) instead of text. This allows the internal state of the odo push process to be more easily consumed by external tools.
+
+Several different types of JSON-formatted events would be output, which correspond to odo container command/action executions:
+- Dev file command execution begun *(command name, start timestamp)*
+- Dev file command execution completed *(error code, end timestamp)*
+- Dev file action execution begun *(action name, parent command name, start timestamp)*
+- Dev file action execution completed *(action name, error code, end timestamp)*
+- Log/console stdout from the actions, one line at a time *(for example, `mvn build` output).* (timestamp)
+  
+(Exact details for which fields are included with events are TBD)
+
+In addition, `odo push -o json` should return a non-zero error code if one of the actions returned a non-zero error code, otherwise zero is returned.
+
+### `odo push -o json` example output
+
+This is what an `odo push -o json` command invocation would look like:
+```
+odo push -o json
+{ "devFileCommandExecutionBegun": { "commandName" : "build", "timestamp" : "(UTC unix epoch seconds.microseconds)" } }
+{ "devFileActionExecutionBegun" : { "commandName" : "build", "actionName" : "run-build-script", "timestamp" : "(...)" } }
+{ "logText" : { "text:" "one line of text received\\n another line of text received", "timestamp" : "(...)" } }
+{ "devFileActionExecutionComplete" : { "errorCode" : 1, ( same as above )} }
+{ "logText" : { "text": (... ), "timestamp" : "(...)" } } # Log text is interleaved with events 
+{ "devFileCommandExecutionComplete": { "success" : false, (same as above) } }
+
+(Exact details on event name, and JSON format are TBD; feedback welcome!)
+```
+
+These events allow an external odo-consuming tool to determine the build status of an application (build succeeded, build failed, build not running).
```

Comment:
Since these are just JSON-formatted strings that are  output to stdout, the consuming tool is free to ignore whichever events it wishes (if event == 'type I don't want' then skip). I don't think there is value add to allowing any additional customization of which events are output by odo.

Reply:
Makes sense :+1:

Reply:
> Thinking out loud: can we make use of k8s health checks for this and/or other status info that we intend to come up with?

K8s health checks appear not to be defined in either odo devfile v1 or v2, so this would be something to consider if/when these are added there. (And these would not apply if odo was running with docker target)